camel/agents/chat_agent.py

```diff
@@ -350,6 +457,94 @@ def handle_stream_response(
         usage_dict = self.get_usage_dict(output_messages, prompt_tokens)
         return output_messages, finish_reasons, usage_dict, response_id
 
+    def step_token_exceed(
+            self, num_tokens: int,
+            called_funcs: List[FunctionCallingRecord]) -> ChatAgentResponse:
+        r"""Return trivial response containing number of tokens and information
+        of called functions when the number of tokens exceeds.
+
+        Args:
+            num_tokens (int): Number of tokens in the messages.
+            called_funcs (List[FunctionCallingRecord]): List of information
+                objects of functions called in the current step.
+
+        Returns:
+            ChatAgentResponse: The struct containing trivial outputs and
+                information about token number and called functions.
+        """
+
+        self.terminated = True
+        output_messages: List[BaseMessage] = []
+
+        info = self.get_info(
+            None,
+            None,
+            ["max_tokens_exceeded"],
+            num_tokens,
+            called_funcs,
+        )
+
+        return ChatAgentResponse(
+            output_messages,
+            self.terminated,
+            info,
+        )
+
+    def step_function_call(
+        self, response: Dict[str, Any]
+    ) -> Tuple[FunctionCallingMessage, FunctionCallingMessage,
+               FunctionCallingRecord]:
+        r"""Execute the function with arguments following the LLM's response.
+
+        Args:
+            response (Dict[str, Any]): the response obtained by calling
+                the LLM model.
+
+        Returns:
+            tuple: a tuple consisting of two obj:`FunctionCallingMessage`,
+                one about the arguments and the other about the execution
+                result, and a struct for logging information about this
+                function call.
+        """
+
+        # Note that when function calling is enabled, `n` is set to 1.
+        choice = response["choices"][0]
+
+        func_name = choice["message"]["function_call"]["name"]
+        func = self.func_dict[func_name]
+
+        args = choice["message"]["function_call"]["arguments"]
+        args = json.loads(args.replace("\'", "\""))
```

Comment:
Just want to understand. Why this replace is needed?

Reply:
@lightaime That is because `json` only supports `"` but not `'` and that's why I firstly implemented using `eval`. But after replacing, `json.loads` can correctly be executed so I just used this.